### PR TITLE
populate file metadata for onnxruntime.dll

### PR DESCRIPTION
**Pull request by orilevari**
**Description**: Populates file information for onnxruntime.dll, follow pattern used currently for inbox winml binaries. ADO builds encode build and source information into version metadata.

**Motivation and Context**
Certain customers as well as inbox shipping require this metadata

### CODE REVIEW

File: cmake/onnxruntime.rc [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// This file REQUIRES the following external definitions:
```
Comment (smk2007) [inline cmake/onnxruntime.rc]: >// [](start = 0, length = 2)

does this file require copyright notice?

# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -823,7 +823,12 @@ else()
   list(APPEND onnxruntime_EXTERNAL_LIBRARIES ${CMAKE_DL_LIBS} Threads::Threads)
 endif()
 
-
+# Version parts for Windows.AI.MachineLearning.dll and onnxruntime.dll
+set(WINML_VERSION_MAJOR_PART   0 CACHE STRING "First part of numeric file/product version.")
+set(WINML_VERSION_MINOR_PART   0 CACHE STRING "Second part of numeric file/product version.")
+set(WINML_VERSION_BUILD_PART   0 CACHE STRING "Third part of numeric file/product version.")
+set(WINML_VERSION_PRIVATE_PART 0 CACHE STRING "Fourth part of numeric file/product version.")
+set(WINML_VERSION_STRING       "Internal Build" CACHE STRING "String representation of file/product version.")
```
Comment (tiagoshibata) [inline cmake/CMakeLists.txt]: We shouldn't have `WINML_VERSION_*` configurations when WinML is disabled... Maybe rename to onnxruntime instead of WINML?

Comment (smk2007) [reply cmake/CMakeLists.txt]: beat me to it!

---
In reply to: [375561884](https://github.com/microsoft/onnxruntime/pull/2978#discussion_r375561884) [](ancestors = 375561884)

File: cmake/onnxruntime.cmake [hunk 3]
```diff
@@ -26,16 +26,29 @@ add_custom_command(OUTPUT ${SYMBOL_FILE} ${CMAKE_CURRENT_BINARY_DIR}/generated_s
...
+    add_library(onnxruntime SHARED 
+      ${SYMBOL_FILE}
+      "${ONNXRUNTIME_ROOT}/core/dll/dllmain.cc"
+      ${PROJECT_SOURCE_DIR}/onnxruntime.rc
+    )
 else()
-    add_library(onnxruntime SHARED ${CMAKE_CURRENT_BINARY_DIR}/generated_source.c)
+    add_library(onnxruntime SHARED 
+      ${CMAKE_CURRENT_BINARY_DIR}/generated_source.c
+      ${PROJECT_SOURCE_DIR}/onnxruntime.rc
```
Comment (tiagoshibata) [inline cmake/onnxruntime.cmake]: Are resources supported outside of WIN32?

File: cmake/onnxruntime.rc [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#define VER_DEBUG 0
+#endif
+
+// -----------------------------------------------------------------------------
+
+VS_VERSION_INFO VERSIONINFO
+FILEVERSION     VER_MAJOR, VER_MINOR, VER_BUILD, VER_PRIVATE
+PRODUCTVERSION  VER_MAJOR, VER_MINOR, VER_BUILD, VER_PRIVATE
+FILEFLAGSMASK   VS_FFI_FILEFLAGSMASK
+FILEFLAGS       VER_DEBUG
```
Comment (smk2007) [inline cmake/onnxruntime.rc]: >FILEFLAGS [](start = 0, length = 9)

Should this file should live in the same directory as the dll source?

File: cmake/onnxruntime.cmake (hunk 3 above)
Comment (smk2007) [reply cmake/onnxruntime.cmake]: they arn't i dont believe there are even apis for accessing them on other platforms.

---
In reply to: [375562814](https://github.com/microsoft/onnxruntime/pull/2978#discussion_r375562814) [](ancestors = 375562814)

File: tools/ci_build/build.py [hunk 5]
```diff
@@ -376,6 +376,33 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
...
+            # MinorPart = 0
+            # BuildPart = 1911
+            # PrivatePart = 123
+            # String = 191101-2300.1.master.0bce7ae
+            cmake_args += ["-DWINML_VERSION_MAJOR_PART=1",
+                        "-DWINML_VERSION_MINOR_PART=0",
+                        f"-DWINML_VERSION_BUILD_PART={YY}{MM}",
+                        f"-DWINML_VERSION_PRIVATE_PART={DD}{HH}",
+                        f"-DWINML_VERSION_STRING=1.0.{build_number}.{source_version[0:7]}"]
+
```
Comment (snnn) [inline tools/ci_build/build.py]: What if they are not defined?

File: cmake/onnxruntime.rc (hunk 1 above)
Comment (orilevari) [reply cmake/onnxruntime.rc]: I followed the lead of winml.rc which did not have a copyright notice. Should I add it to both files?

File: tools/ci_build/build.py (hunk 5 above)
Comment (orilevari) [reply tools/ci_build/build.py]: The version variables will be set to default values of 0.0.0.0 and "Internal Build" for builds that aren't done through ADO.

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (orilevari) [reply cmake/CMakeLists.txt]: I was trying to differentiate it from the other ORT_VERSION variable that is extracted from the VERSION_NUMBER file but not used for populating the ORT metadata. How about just VERSION_*?

File: tools/ci_build/build.py [hunk 6]
```diff
@@ -376,6 +376,33 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
...
+            DD = matches.group(3)
+            HH = matches.group(4)
+
+            # Example (BuildNumber: 191101-2300.1.master, SourceVersion: 0bce7ae6755c792eda558e5d27ded701707dc404)
+            # MajorPart = 1
+            # MinorPart = 0
+            # BuildPart = 1911
+            # PrivatePart = 123
+            # String = 191101-2300.1.master.0bce7ae
+            cmake_args += ["-DWINML_VERSION_MAJOR_PART=1",
```
Comment (snnn) [inline tools/ci_build/build.py]: Would it be better to get the major and minor part from https://github.com/microsoft/onnxruntime/blob/master/VERSION_NUMBER ?

File: cmake/onnxruntime.cmake (hunk 3 above)
Comment (orilevari) [reply cmake/onnxruntime.cmake]: ah I though WIN32 was filtering x86, but now I see it is filtering any Windows build. I'll fix that.

File: cmake/onnxruntime.rc (hunk 1 above)
Comment (pranavsharma) [reply cmake/onnxruntime.rc]: Yes, please.

File: tools/ci_build/build.py (hunk 6 above)
Comment (orilevari) [reply tools/ci_build/build.py]: For context, @pranavsharma mentioned via email that it is okay for the major and minor part to match that of the winml binary. It is also useful of the two to match for auditing purposes, but that being said the build information would be enough information to match up the dlls. I'll go ahead and change this.

File: cmake/onnxruntime.rc (hunk 4 above)
Comment (orilevari) [reply cmake/onnxruntime.rc]: makes sense, I went ahead and moved it there

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (orilevari) [reply cmake/CMakeLists.txt]: Ended up specifying ORT or WINML for variables that are specific to one or the other, and include no prefix for shared variables

File: tools/ci_build/build.py [hunk 7]
```diff
@@ -376,6 +376,42 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
     if is_windows():
         cmake_args += cmake_extra_args
 
+    # ADO pipelines will store the pipeline build number (e.g. 191101-2300.1.master) and 
+    # source version in environment variables. If present, use these values to define the 
+    # WinML/DML DLL versions.
+    build_number = os.getenv('Build_BuildNumber')
+    source_version = os.getenv('Build_SourceVersion')
+    if build_number and source_version:
+        build_matches = re.match(r"^(\d\d)(\d\d)(\d\d)-(\d\d)(\d\d)\.(\d)\.(\S+)$", build_number)
```
Comment (tiagoshibata) [inline tools/ci_build/build.py]: Nit: `re.fullmatch` does the same as ^$

File: tools/ci_build/build.py [hunk 8]
```diff
@@ -376,6 +376,42 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
...
+                ort_major = ort_version_matches.group(1)
+                ort_minor = ort_version_matches.group(2)
+                # Example (BuildNumber: 191101-2300.1.master, SourceVersion: 0bce7ae6755c792eda558e5d27ded701707dc404)
+                # MajorPart = 1
+                # MinorPart = 0
+                # BuildPart = 1911
+                # PrivatePart = 123
+                # String = 191101-2300.1.master.0bce7ae
+                cmake_args += ["-DWINML_VERSION_MAJOR_PART=1",
+                            "-DWINML_VERSION_MINOR_PART=0",
```
Comment (smk2007) [inline tools/ci_build/build.py]: Can we get rid of these and just use the ORT major/minor in winml?

Comment (orilevari) [reply tools/ci_build/build.py]: done!

File: tools/ci_build/build.py [hunk 9]
```diff
@@ -376,6 +376,39 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
     if is_windows():
         cmake_args += cmake_extra_args
 
+    # ADO pipelines will store the pipeline build number (e.g. 191101-2300.1.master) and 
+    # source version in environment variables. If present, use these values to define the 
+    # WinML/DML DLL versions.
```
Comment (smk2007) [inline tools/ci_build/build.py]: >DML DLL [](start = 12, length = 7)

not dml!

Comment (orilevari) [reply tools/ci_build/build.py]: oh whoops!

**Pull request merged**